pkg/operator/resource/resourceapply/monitoring.go

```diff
@@ -2,148 +2,150 @@ package resourceapply
 	}
-
-	if !modified {
-		return nil, false, nil
-	}
-
-	if klog.V(2).Enabled() {
-		klog.Infof("ServiceMonitor %q changes: %v", namespace+"/"+required.GetName(), JSONPatchNoError(existing, toUpdate))
+	if !found {
+		return nil, false, errorsstdlib.New(fmt.Sprintf("metadata not found in %s", existingCopy.GetName()))
 	}
-
-	newObj, err := client.Resource(serviceMonitorGVR).Namespace(namespace).Update(ctx, toUpdate, metav1.UpdateOptions{})
+	requiredObjectMeta, found, err := unstructured.NestedMap(required.Object, "metadata")
 	if err != nil {
-		recorder.Warningf("ServiceMonitorUpdateFailed", "Failed to update ServiceMonitor.monitoring.coreos.com/v1: %v", err)
-		return nil, true, err
+		return nil, false, err
+	}
+	if !found {
+		return nil, false, errorsstdlib.New(fmt.Sprintf("metadata not found in %s", required.GetName()))
 	}
 
-	recorder.Eventf("ServiceMonitorUpdated", "Updated ServiceMonitor.monitoring.coreos.com/v1 because it changed")
-	return newObj, true, err
-}
-
-var prometheusRuleGVR = schema.GroupVersionResource{Group: "monitoring.coreos.com", Version: "v1", Resource: "prometheusrules"}
-
-// ApplyPrometheusRule applies the PrometheusRule
-func ApplyPrometheusRule(ctx context.Context, client dynamic.Interface, recorder events.Recorder, required *unstructured.Unstructured) (*unstructured.Unstructured, bool, error) {
-	namespace := required.GetNamespace()
-
-	existing, err := client.Resource(prometheusRuleGVR).Namespace(namespace).Get(ctx, required.GetName(), metav1.GetOptions{})
-	if errors.IsNotFound(err) {
-		newObj, createErr := client.Resource(prometheusRuleGVR).Namespace(namespace).Create(ctx, required, metav1.CreateOptions{})
-		if createErr != nil {
-			recorder.Warningf("PrometheusRuleCreateFailed", "Failed to create PrometheusRule.monitoring.coreos.com/v1: %v", createErr)
-			return nil, true, createErr
-		}
-		recorder.Eventf("PrometheusRuleCreated", "Created PrometheusRule.monitoring.coreos.com/v1 because it was missing")
-		return newObj, true, nil
+	// Cast the metadata to the correct type.
+	var existingObjectMetaTyped, requiredObjectMetaTyped metav1.ObjectMeta
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(existingObjectMeta, &existingObjectMetaTyped)
+	if err != nil {
+		return nil, false, err
 	}
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(requiredObjectMeta, &requiredObjectMetaTyped)
 	if err != nil {
 		return nil, false, err
 	}
 
-	existingCopy := existing.DeepCopy()
+	// Check if the metadata objects differ.
+	didMetadataModify := resourcemerge.BoolPtr(false)
+	resourcemerge.EnsureObjectMeta(didMetadataModify, &existingObjectMetaTyped, requiredObjectMetaTyped)
 
-	toUpdate, modified, err := ensureGenericSpec(required, existingCopy, noDefaulting, equality.Semantic)
+	// Deep-check the spec objects for equality, and update the cache in either case.
+	existingCopy, didSpecModify, err := ensureGenericSpec(required, existingCopy, noDefaulting, equality.Semantic)
```

Comment:
IIUC, the default values set by the server (that are not overwritten) are ignored, thus in most cases, it will consider the objects as not equal.

Do we plan to allow overwriting `noDefaulting` and  `equality.Semantic` in the future?
In the meanwhile I think we should state it explicitly (as "improved" in `XXXImproved` may be misleading): "if you want to make sure UPDATES are only triggered when needed, you should enable the cache."

CMO doesn't use the cache for example (for now), as the no-cache `ApplyXXX` utils it's using now ensure limited UPDATES even without the cache.

Reply:
> Do we plan to allow overwriting noDefaulting and  equality.Semantic in the future?

I'm +1 on moving `noDefaulting` and `equality.Semantic` parameters to the function signature.

> In the meanwhile I think we should state it explicitly (as "improved" in XXXImproved may be misleading): "if you want to make sure UPDATES are only triggered when needed, you should enable the cache."

I'm not sure what is being referred to here? Are you suggesting a comment?

> CMO doesn't use the cache for example (for now), as the no-cache ApplyXXX utils it's using now ensure limited UPDATES even without the cache.

I believe we will need to initialize a cache for CMO to be able to benefit from the changes being made here, i.e., dropping `ApplyXXXImproved` in favor of `ApplyUnstructuredResourceImproved`.

Reply:
Friendly ping. :)

Reply:
Additionally, we might want to supply custom a `equalityChecker` or `defaultingFunc` that caters to the CRD to make things more robust.

Reply:
@machine424 suggested we drop these parameters, since we will only be introducing this function for the Prometheus object for now. If there are any defaults that need to be set for it, they should be set in Prometheus operator, which makes sense.

However, when and if we decide to expand this in the future, we will need to add these to cater to other components' use cases that don't have such middlewares (and require setting defaults in CMO itself, as that's where their configuration is defined).